extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,298 @@
+   */
+  constructor(filter, opt_allowNull) {
+    /** @type {!function(...?):(?|!Promise<?>)} */
+    this.filter = filter;
+
+    /** @type{boolean} */
+    this.allowNull = !!opt_allowNull;
+  }
+}
+
+
+/**
+ * The structure that contains all details needed to expand a template
+ * @struct
+ * @const
+ * @package For type.
+ */
+export class ExpansionOptions {
+  /**
+   * @param {!Object<string, *>} vars
+   * @param {number=} opt_iterations
+   * @param {boolean=} opt_noEncode
+   */
+  constructor(vars, opt_iterations, opt_noEncode) {
+    /** @const {!Object<string, string|Array<string>>} */
+    this.vars = vars;
+    /** @const {!number} */
+    this.iterations = opt_iterations === undefined ? 2 : opt_iterations;
+    /** @const {boolean} */
+    this.noEncode = !!opt_noEncode;
+  }
+}
+
+
+
+/**
+ * @param {!string} str
+ * @param {Number} s
+ * @param {Number} l
+ * @return {string}
+ */
+function substrFilter(str, s, l) {
+  const start = Number(s);
+  let length = str.length;
+  user().assert(isFiniteNumber(start),
+      'Start index ' + start + 'in substr filter should be a number');
+  if (l) {
+    length = Number(l);
+    user().assert(isFiniteNumber(length),
+        'Length ' + length + ' in substr filter should be a number');
+  }
+
+  return str.substr(start, length);
+}
+
+/**
+ * @param {*} value
+ * @param {*} defaultValue
+ * @return {*}
+ */
+function defaultFilter(value, defaultValue) {
+  return value || user().assertString(defaultValue);
+}
+
+
+export class VariableService {
+  /**
+   * @param {!Window} window
+   */
+  constructor(window) {
+
+    /** @private {!Window} */
+    this.win_ = window;
+
+    /** @private {!Object<string, FilterDef>} */
+    this.filters_ = map();
+
+    this.register_('default', new FilterDef(defaultFilter, true));
+    this.register_('substr', new FilterDef(substrFilter));
+    this.register_('trim', new FilterDef(
+        value => user().assertString(value).trim()));
+    this.register_('json', new FilterDef(value => JSON.stringify(value)));
+    this.register_('toLowerCase', new FilterDef(value =>
+        user().assertString(value).toLowerCase()));
+    this.register_('toUpperCase', new FilterDef(value =>
+        user().assertString(value).toUpperCase()));
+    this.register_('not', new FilterDef(value => String(!value)));
+    this.register_('base64', new FilterDef(
+        value => btoa(user().assertString(value))));
+    this.register_('hash', new FilterDef(this.hashFilter_.bind(this)));
+    this.register_('if', new FilterDef(
+        (value, thenValue, elseValue) => value ? thenValue : elseValue, true));
+  }
+
+  /**
+   * @param {string} name
+   * @param {FilterDef} filter
+   */
+  register_(name, filter) {
+    dev().assert(!this.filters_[name], 'Filter "' + name
+        + '" already registered.');
+    this.filters_[name] = filter;
+  }
+
+  /**
+   * @param {string} filterStr
+   * @return {Object<string, FilterDef|Array<string>>}
```

Comment:
Had to use `@return {?{filter: !Filter, args: !Array<string>}}` as I return null when the filter is not found.